### DISCUSSION TITLE
Problem with Console and Kiauh shell_command Extension

### DISCUSSION THREAD

**Discussion by Dalinaffeti** [mainsail-crew/mainsail#1769] [Help] [answered]
Hi i am new to Mainsail. I want to launch a python script with the shell_command extension of Kiauh. But with simple tests like 'echo hello world' i have no outputs on the console. In the screenshot below you can find what i am talking about. Every help will be appreciate.

The problem is i dont know which is causing this is it the Kiauh or the MainsailOS
![gitkiauh](https://github.com/mainsail-crew/mainsail/assets/72552658/d922d6fc-0757-4ce9-b4a9-5c5afb0721a7)

**Comment by Dalinaffeti** [comment]
i forgot to add a screenshot of the content of my shell_command file:
![Screenshot (60)](https://github.com/mainsail-crew/mainsail/assets/72552658/0bf3ff7d-4f62-427e-bc65-ee97f90abe50)

**Comment by meteyou** [comment]
An echo will not be displayed in the console because the output has nothing to do with the Klipper console.

And it's not recommended to use sudo commands with it...

**Reply by Dalinaffeti** [reply]
did you use the kiauh shell_command extension with mainsail before?

**Reply by meteyou** [reply]
yes

**Reply by meteyou** [reply]
(and it is not an extension... is more like a "hack". If it is possible without this "module", I would not use it...)

**Reply by Dalinaffeti** [reply]
because i followed a youtube video on how to use it and in the video when gives a makro like hello_world it will be shown on the console.

It is shown at 5:11
https://www.youtube.com/watch?v=qkn3Su7QCpk&t=313s

**Comment by Dalinaffeti** [comment]
> because i followed a youtube video on how to use it and in the video when gives a makro like hello_world it will be shown on the console.
> 
> It is shown at 5:11 https://www.youtube.com/watch?v=qkn3Su7QCpk&t=313s
unfortunately i didnt found examples on how it works

**Reply by meteyou** [reply]
I checked the video, and I didn't know that the output would be displayed in the Klipper console.

the only issue I found is the space between cmd= and the gcode command name

**Reply by Dalinaffeti** [reply]
i removed it but still the same

**Reply by meteyou** [reply]
pls post a klippy.log

**Reply by Dalinaffeti** [reply]
sorry for the belated response i didnt had my raspberry pi with me. The latest klippy.log that i found dated 13.01.2024. It is a long file i wasnt allowed tro share it on github do you have any spesific thing or part that i should share?

**Reply by Dalinaffeti** [reply]
jut to add something to it. This is an image thet colleague did and i installed it on my raspberyy pi, so that i work with the shell commands. In the config file i put: 
[mcu host]
serial: /tmp/klipper_host_mcu
So that i could work around with it

**Reply by meteyou** [reply]
you can use the "rollover logs" button in mainsail to start a new logfile. so you have a short one. it is the button on the right top of the "logfiles panel":
![image](https://github.com/mainsail-crew/mainsail/assets/8167632/5cf6d083-c385-40a2-a2d6-2b50484ce102)

the sheet with the refresh symbol.

**Reply by Dalinaffeti** [reply]
Starting Klippy...
Args: ['/home/pi/klipper/klippy/klippy.py', '/home/pi/printer_data/config/printer.cfg', '-l', '/home/pi/printer_data/logs/klippy.log', '-I', '/home/pi/printer_data/comms/klippy.serial', '-a', '/home/pi/printer_data/comms/klippy.sock']
Git version: 'v0.11.0-299-gb1f597c5-dirty'
Untracked files: klippy/extras/gcode_shell_command.py
Branch: master
Remote: origin
Tracked URL: https://github.com/Klipper3d/klipper.git
CPU: 4 core ?
Python: '3.9.2 (default, Mar 12 2021, 04:06:34) \n[GCC 10.2.1 20210110]'
Start printer at Sun Jan 14 02:27:31 2024 (1705195651.8 634.7)
===== Config file =====
[gcode_shell_command hello_world]
command = echo hello worlders
timeout = 2.
verbose = True

[gcode_macro HELLO_WORLD]
gcode = M978

[gcode_shell_command open_laser]
command = sudo python /home/pi/led_blink.py
verbose = True

[gcode_shell_command StartLaser]
command = SERIAL =$(python -c 'import  /config/Python Scripts/LaserControl; LaserControl.StartLaser()')
timeout = 2.
verbose = True

[gcode_macro START_LASER]
gcode = 
	RUN_SHELL_COMMAND CMD=StartLaser

[gcode_macro HELLO_WORLDD]
gcode = 
	RUN_SHELL_COMMAND CMD=hello_world

[gcode_shell_command close_laser]
command = sudo ^C /home/pi/led_blink.py
verbose = True

[virtual_sdcard]
path = ~/printer_data/gcodes
on_error_gcode = CANCEL_PRINT

[pause_resume]

[display_status]

[respond]

[gcode_macro CANCEL_PRINT]
description = Cancel the actual running print
rename_existing = CANCEL_PRINT_BASE
gcode = 
	
	{% set client = printer['gcode_macro _CLIENT_VARIABLE']|default({}) %}
	{% set allow_park = client.park_at_cancel|default(false)|lower == 'true' %}
	{% set retract = client.cancel_retract|default(5.0)|abs %}
	
	{% set park_x = "" if (client.park_at_cancel_x|default(none) is none)
	else "X=" ~ client.park_at_cancel_x %}
	{% set park_y = "" if (client.park_at_cancel_y|default(none) is none)
	else "Y=" ~ client.park_at_cancel_y %}
	{% set custom_park = park_x|length > 0 or park_y|length > 0 %}
	
	
	{% if printer['gcode_macro PAUSE'].restore_idle_timeout > 0 %}
	SET_IDLE_TIMEOUT TIMEOUT={printer['gcode_macro PAUSE'].restore_idle_timeout}
	{% endif %}
	{% if (custom_park or not printer.pause_resume.is_paused) and allow_park %} _TOOLHEAD_PARK_PAUSE_CANCEL {park_x} {park_y} {% endif %}
	_CLIENT_RETRACT LENGTH={retract}
	TURN_OFF_HEATERS
	M106 S0
	
	SET_PAUSE_NEXT_LAYER ENABLE=0
	SET_PAUSE_AT_LAYER ENABLE=0 LAYER=0
	CANCEL_PRINT_BASE

[gcode_macro PAUSE]
description = Pause the actual running print
rename_existing = PAUSE_BASE
variable_restore_idle_timeout = 0
gcode = 
	PAUSE_BASE
	_TOOLHEAD_PARK_PAUSE_CANCEL

[gcode_macro RESUME]
description = Resume the actual running print
rename_existing = RESUME_BASE
variable_last_extruder_temp = {'restore': False, 'temp': 0}
gcode = 
	
	{% set client = printer['gcode_macro _CLIENT_VARIABLE']|default({}) %}
	{% set velocity = printer.configfile.settings.pause_resume.recover_velocity %}
	{% set sp_move = client.speed_move|default(velocity) %}
	
	
	{% if printer['gcode_macro PAUSE'].restore_idle_timeout > 0 %}
	SET_IDLE_TIMEOUT TIMEOUT={printer['gcode_macro PAUSE'].restore_idle_timeout}
	{% endif %}
	{% if printer.idle_timeout.state|upper == "IDLE" %}
	{% if last_extruder_temp.restore %} M109 S{last_extruder_temp.temp} {% endif %}
	{% endif %}
	_CLIENT_EXTRUDE
	RESUME_BASE VELOCITY={params.VELOCITY|default(sp_move)}

[gcode_macro SET_PAUSE_NEXT_LAYER]
description = Enable a pause if the next layer is reached
gcode = 
	{% set pause_next_layer = printer['gcode_macro SET_PRINT_STATS_INFO'].pause_next_layer %}
	{% set ENABLE = params.ENABLE|default(1)|int != 0 %}
	{% set MACRO = params.MACRO|default(pause_next_layer.call, True) %}
	SET_GCODE_VARIABLE MACRO=SET_PRINT_STATS_INFO VARIABLE=pause_next_layer VALUE="{{ 'enable': ENABLE, 'call': MACRO }}"

[gcode_macro SET_PAUSE_AT_LAYER]
description = Enable/disable a pause if a given layer number is reached
gcode = 
	{% set pause_at_layer = printer['gcode_macro SET_PRINT_STATS_INFO'].pause_at_layer %}
	{% set ENABLE = params.ENABLE|int != 0 if params.ENABLE is defined
	else params.LAYER is defined %}
	{% set LAYER = params.LAYER|default(pause_at_layer.layer)|int %}
	{% set MACRO = params.MACRO|default(pause_at_layer.call, True) %}
	SET_GCODE_VARIABLE MACRO=SET_PRINT_STATS_INFO VARIABLE=pause_at_layer VALUE="{{ 'enable': ENABLE, 'layer': LAYER, 'call': MACRO }}"

[gcode_macro SET_PRINT_STATS_INFO]
rename_existing = SET_PRINT_STATS_INFO_BASE
description = Overwrite, to get pause_next_layer and pause_at_layer feature
variable_pause_next_layer = { 'enable': False, 'call': "PAUSE" }
variable_pause_at_layer = { 'enable': False, 'layer': 0, 'call': "PAUSE" }
gcode = 
	{% if pause_next_layer.enable %}
	RESPOND TYPE=echo MSG='{"%s, forced by pause_next_layer" % pause_next_layer.call}'
	{pause_next_layer.call}
	SET_PAUSE_NEXT_LAYER ENABLE=0
	{% elif pause_at_layer.enable and params.CURRENT_LAYER is defined and params.CURRENT_LAYER|int == pause_at_layer.layer %}
	RESPOND TYPE=echo MSG='{"%s, forced by pause_at_layer [%d]" % (pause_at_layer.call, pause_at_layer.layer)}'
	{pause_at_layer.call}
	SET_PAUSE_AT_LAYER ENABLE=0
	{% endif %}
	SET_PRINT_STATS_INFO_BASE {rawparams}

[gcode_macro _TOOLHEAD_PARK_PAUSE_CANCEL]
description = Helper: park toolhead used in PAUSE and CANCEL_PRINT
gcode = 
	
	{% set client = printer['gcode_macro _CLIENT_VARIABLE']|default({}) %}
	{% set velocity = printer.configfile.settings.pause_resume.recover_velocity %}
	{% set use_custom     = client.use_custom_pos|default(false)|lower == 'true' %}
	{% set custom_park_x  = client.custom_park_x|default(0.0) %}
	{% set custom_park_y  = client.custom_park_y|default(0.0) %}
	{% set park_dz        = client.custom_park_dz|default(2.0)|abs %}
	{% set sp_hop         = client.speed_hop|default(15) * 60 %}
	{% set sp_move        = client.speed_move|default(velocity) * 60 %}
	
	{% set origin    = printer.gcode_move.homing_origin %}
	{% set act       = printer.gcode_move.gcode_position %}
	{% set max       = printer.toolhead.axis_maximum %}
	{% set cone      = printer.toolhead.cone_start_z|default(max.z) %}
	{% set round_bed = True if printer.configfile.settings.printer.kinematics is in ['delta','polar','rotary_delta','winch']
	else False %}
	
	{% set z_min = params.Z_MIN|default(0)|float %}
	{% set z_park = [[(act.z + park_dz), z_min]|max, (max.z - origin.z)]|min %}
	{% set x_park = params.X       if params.X is defined
	else custom_park_x  if use_custom
	else 0.0            if round_bed
	else (max.x - 5.0) %}
	{% set y_park = params.Y       if params.Y is defined
	else custom_park_y  if use_custom
	else (max.y - 5.0)  if round_bed and z_park < cone
	else 0.0            if round_bed
	else (max.y - 5.0) %}
	
	_CLIENT_RETRACT
	{% if "xyz" in printer.toolhead.homed_axes %}
	G90
	G1 Z{z_park} F{sp_hop}
	G1 X{x_park} Y{y_park} F{sp_move}
	{% if not printer.gcode_move.absolute_coordinates %} G91 {% endif %}
	{% else %}
	RESPOND TYPE=echo MSG='Printer not homed'
	{% endif %}

[gcode_macro _CLIENT_EXTRUDE]
description = Extrudes, if the extruder is hot enough
gcode = 
	
	{% set client = printer['gcode_macro _CLIENT_VARIABLE']|default({}) %}
	{% set use_fw_retract = (client.use_fw_retract|default(false)|lower == 'true') and (printer.firmware_retraction is defined) %}
	{% set length = params.LENGTH|default(client.unretract)|default(1.0)|float %}
	{% set speed = params.SPEED|default(client.speed_unretract)|default(35) %}
	{% set absolute_extrude = printer.gcode_move.absolute_extrude %}
	
	{% if printer.toolhead.extruder != '' %}
	{% if printer[printer.toolhead.extruder].can_extrude %}
	{% if use_fw_retract %}
	{% if length < 0 %}
	G10
	{% else %}
	G11
	{% endif %}
	{% else %}
	M83
	G1 E{length} F{(speed|float|abs) * 60}
	{% if absolute_extrude %}
	M82
	{% endif %}
	{% endif %}
	{% else %}
	RESPOND TYPE=echo MSG='Extruder not hot enough'
	{% endif %}
	{% endif %}

[gcode_macro _CLIENT_RETRACT]
description = Retracts, if the extruder is hot enough
gcode = 
	{% set client = printer['gcode_macro _CLIENT_VARIABLE']|default({}) %}
	{% set length = params.LENGTH|default(client.retract)|default(1.0)|float %}
	{% set speed = params.SPEED|default(client.speed_retract)|default(35) %}
	
	_CLIENT_EXTRUDE LENGTH=-{length|float|abs} SPEED={speed|float|abs}

[mcu]
serial = /tmp/klipper_host_mcu

[mcu host]
serial = /tmp/klipper_host_mcu

[printer]
kinematics = cartesian
max_velocity = 140.0
max_accel = 1000
square_corner_velocity = 4.0
max_z_velocity = 20
max_z_accel = 500

[stepper_x]
step_pin = PE2
dir_pin = PE1
enable_pin = !PE3
microsteps = 16
rotation_distance = 20
endstop_pin = ^!PC1
position_endstop = 360
position_min = 0
position_max = 360
homing_speed = 10

[stepper_y]
step_pin = PD5
dir_pin = PD4
enable_pin = !PD6
microsteps = 16
rotation_distance = 20
endstop_pin = ^!PC3
position_endstop = 360
position_min = 0
position_max = 360
homing_speed = 10

[stepper_y1]
step_pin = PD15
dir_pin = PD14
enable_pin = !PC7
microsteps = 16
rotation_distance = 20

[stepper_z]
step_pin = PA15
dir_pin = PA8
enable_pin = !PD1
microsteps = 16
rotation_distance = 8
endstop_pin = ^!PC0
position_endstop = 350
position_min = 0
position_max = 360
homing_speed = 10

[fan]
pin = PB7

[output_pin motor_power]
pin = PC13
value = 1

[tmc2208 stepper_x]
uart_pin = PE0
run_current = 0.800

[tmc2208 stepper_y]
uart_pin = PD3
run_current = 0.800

[tmc2208 stepper_y1]
uart_pin = PC6
run_current = 0.800

[tmc2208 stepper_z]
uart_pin = PD0
run_current = 0.800

[gcode_macro pwm_pulver_pwm_on]
gcode = 
	M42 P18 S{printer.pwm_pulver.default_parameter_pulver_pwm}
	G4 P{toolhead.injection_pulver_time}

[gcode_macro pwm_pulver_pwm_off]
gcode = 
	M42 P18 S0

[gcode_macro Laser_on]
gcode = SET_PIN PIN=Laser VALUE=1

[gcode_macro Laser_off]
gcode = SET_PIN PIN=Laser VALUE=0

[output_pin Laser]
pin = host:gpio26
value = 0

[gcode_macro Pulver_on]
gcode = SET_PIN PIN=Pulver VALUE=0.80

[gcode_macro Pulver_off]
gcode = SET_PIN PIN=Pulver VALUE=0

[output_pin Pulver]
pin = host:gpio16
pwm = true
value = 0
cycle_time = 0.00143
=======================
mcu 'mcu': Starting connect
webhooks client 4123667016: New connection
mcu 'mcu': got {'count': 139, 'sum': 160250, 'sumsq': 988634, '#name': 'stats', '#sent_time': 635.129905788, '#receive_time': 635.130020214}
webhooks client 4123667016: Client info {'program': 'Moonraker', 'version': 'v0.8.0-186-g2641fc5'}
Loaded MCU 'mcu' 118 commands (v0.11.0-297-g5edc7fee / gcc: (Raspbian 10.2.1-6+rpi1) 10.2.1 20210110 binutils: (GNU Binutils for Raspbian) 2.35.2)
MCU 'mcu' config: ADC_MAX=4095 CLOCK_FREQ=50000000 MCU=linux PCA9685_MAX=4096 PWM_MAX=32768 STATS_SUMSQ_BASE=256
mcu 'host': Starting connect
b'Got EOF when reading from device'

**Reply by meteyou** [reply]
pls upload it as a file and don't copy&paste the content of it...

**Reply by Dalinaffeti** [reply]
[klippy (3).log](https://github.com/mainsail-crew/mainsail/files/14343168/klippy.3.log)

**Reply by meteyou** [reply]
Klipper is not starting in this log. do you have a log with a working klipper?

**Reply by Dalinaffeti** [reply]
yes it is from my colleague which i got the dsic image from. For him everything is working
[klippy (4).log](https://github.com/mainsail-crew/mainsail/files/14360490/klippy.4.log)

**Reply by meteyou** [reply]
I see many times these error lines:
```
Unknown command:"/"
Unknown command:"/"
Unknown command:"/8"
```

Why do you want to use Python scripts to enable the GPIO pins for your laser?

**Reply by Dalinaffeti** [reply]
my Laser uses the RS232C and not G-code therefore i need to manage seperately. Maybe i forgot to mention something: i dont have a printer connected to my RPI. I wanted to do the testing with the shell_command at first and then connect everything together

**Reply by meteyou** [reply]
ok. let me test today evening something.

**Reply by Dalinaffeti** [reply]
i solved the console and the shell_command thing :)

**Reply by meteyou** [reply]
what was the issue?

**Reply by Dalinaffeti** [reply, marked as answer]
it was the mcu. I flashed the firmware to use the raspberryPi as an mcu

**Reply by meteyou** [reply]
yes. Klipper needs a MCU to start. without any MCU, it will not execute any Gcode.
